process/asyncExecution/executionManager/executionManager.go

```diff
@@ -0,0 +1,217 @@
+func (em *executionManager) AddPairForExecution(pair queue.HeaderBodyPair) error {
+	// lock the internal mutex to avoid any concurrent removal requests
+	em.mut.Lock()
+	defer em.mut.Unlock()
+
+	return em.blocksQueue.AddOrReplace(pair)
+}
+
+// GetPendingExecutionResults calls the same method from executionResultsTracker
+func (em *executionManager) GetPendingExecutionResults() ([]data.BaseExecutionResultHandler, error) {
+	return em.executionResultsTracker.GetPendingExecutionResults()
+}
+
+// SetLastNotarizedResult calls the same method from executionResultsTracker
+func (em *executionManager) SetLastNotarizedResult(executionResult data.BaseExecutionResultHandler) error {
+	return em.executionResultsTracker.SetLastNotarizedResult(executionResult)
+}
+
+// CleanConfirmedExecutionResults calls the same method from executionResultsTracker
+func (em *executionManager) CleanConfirmedExecutionResults(header data.HeaderHandler) error {
+	return em.executionResultsTracker.CleanConfirmedExecutionResults(header)
+}
+
+// RemoveAtNonceAndHigher removes the header-body pair at the specified nonce
+// and all pairs with higher nonces from the queue.
+// if anything was processed after that nonce, remove the execution results
+// and rollback blockchain state
+func (em *executionManager) RemoveAtNonceAndHigher(nonce uint64) error {
+	em.mut.Lock()
+	defer em.mut.Unlock()
+
+	lastNotarizedResult, err := em.executionResultsTracker.GetLastNotarizedExecutionResult()
+	if err != nil {
+		return err
+	}
+
+	lastNotarizedNonce := lastNotarizedResult.GetHeaderNonce()
+	nonceToRemove := nonce
+	if nonce <= lastNotarizedNonce {
+		nonceToRemove = lastNotarizedNonce + 1
+		log.Debug("attempting to remove nonce lower than last notarized execution result, updating it",
+			"nonce provided", nonce,
+			"last notarized nonce", lastNotarizedNonce,
+			"removing from nonce", nonceToRemove,
+		)
+	}
+
+	// first pause the headers executor which will block it from popping new headers
+	// but allow it to finish anything currently processing
+	em.headersExecutor.PauseExecution()
+
+	// remove from queue
+	removedNonces := em.blocksQueue.RemoveAtNonceAndHigher(nonceToRemove)
+	if len(removedNonces) > 0 && removedNonces[0] == nonceToRemove {
+		// if the first nonce removed is the initial one,
+		// it means it was still in queue and was not processed.
+		// no matter how many were removed, safe to resume execution
+		em.headersExecutor.ResumeExecution()
+
+		return nil
+	}
+
+	// if the initial nonce was not returned as removed from the queue,
+	// it means that it was already popped for execution (and perhaps not the only one).
+	// inform executionResultsTracker to remove all nonces >= than the provided one
+	err = em.executionResultsTracker.RemoveFromNonce(nonceToRemove)
+	if err != nil {
+		return err
+	}
+
+	// update blockchain with the last executed header, similar to headersExecution
+	err = em.updateBlockchainAfterRemoval(lastNotarizedResult)
+	if err != nil {
```

Comment:
maybe in this kind of cases, we leave the option for the manager to reset the components so we can start over.
E.g. we implement a reset method that re-initializes all sub-components (including the call for resume)

Reply:
added a todo